resteasy/src/main/java/com/linecorp/armeria/client/resteasy/ArmeriaJaxrsClientEngine.java

```diff
@@ -0,0 +1,306 @@
+                    = new ByteBufferBackedOutputStream(bufferSize, buff -> {
+                        if (buff.hasRemaining()) {
+                            requestWriter.write(HttpData.wrap(Unpooled.wrappedBuffer(buff)));
+                        }
+                    });
+            request.getDelegatingOutputStream().setDelegate(requestContentStream);
+            try {
+                request.writeRequestBody(request.getEntityStream());
+                requestContentStream.close();
+            } catch (IOException e) {
+                requestWriter.close(e);
+                throw new RuntimeException("Failed to write the request body!", e);
+            }
+            requestWriter.close();
+            return asyncResponse;
+        }
+    }
+
+    private <T> CompletableFuture<T> handleAsyncResponse(ClientConfiguration clientConfiguration,
+                                                         HttpResponse asyncResponse,
+                                                         boolean buffered,
+                                                         ResultExtractor<T> extractor) {
+        if (buffered) {
+            // Request + Response fully buffered in memory. Future signalled after the entire response
+            // received and processed.
+            final CompletableFuture<AggregatedHttpResponse> responseFuture = asyncResponse.aggregate();
+            return responseFuture.thenApply(aggregatedResponse -> {
+                final HttpData responseContent = aggregatedResponse.content();
+                final ClientResponse response = new ResteasyClientResponseImpl(clientConfiguration,
+                                                                               aggregatedResponse.headers(),
+                                                                               responseContent.toInputStream());
+                return extractor.extractResult(response);
+            });
+        } else {
+            // Unbuffered: Future returns immediately after headers. Reading the response-stream blocks,
+            // but one may check InputStream#available() to prevent blocking.
+            final HttpMessageStream messageStream =
+                    readTimeout == null ? HttpMessageStream.of(asyncResponse)
+                                        : HttpMessageStream.of(asyncResponse, readTimeout);
+            return messageStream.awaitHeaders()
+                                .thenApply(asyncResponseHeaders -> {
+                                    final ClientResponse response =
+                                            new ResteasyClientResponseImpl(clientConfiguration,
+                                                                           asyncResponseHeaders,
+                                                                           messageStream.content());
+                                    return extractor.extractResult(response);
+                                });
```

Comment:
- Yes, as the underlying Netty operates with buffered IO we will always have some kind of "buffering" inside. What RESTEasy call **buffered** is when request or response buffered/aggregated **in full**!
- No, I don't think so. For response processing, this `headersFuture` will complete when the headers portion received (for the request processing it will be triggered right away). The headers portion required to construct an instance of `ClientResponse`.

Reply:
> No, I don't think so. For response processing, this headersFuture will complete when the headers portion received (for the request processing it will be triggered right away). The headers portion required to construct an instance of ClientResponse.

I misunderstood. Then how about using `HttpResponse.split()` to get a headers future?
```java
final SplitHttpResponse splitResponse = asyncResponse.split();
return splitResponse.headers()
        .thenApply(headers -> {
            splitResponse.body()... ?
        });
```

Reply:
I think I've responded to this already...
As you can see, the implementations of `DefaultSplitHttpResponse` and `HttpMessageStream` are very similar in respect to handling the `headersFuture`. Since I'm implementing `Subscriber` interface and I need to handle both `HttpRequest` and `HttpResponse` in a consistent fashion, I don't see much value of using `HttpResponse.split()`.

Reply:
Sorry, I missed your previous comment. 😅  

>  I'm implementing Subscriber interface and I need to handle both HttpRequest and HttpResponse in a consistent fashion

That makes sense. Let's keep it as it is.